### ISSUE TITLE
Live fiber dumps

### ISSUE THREAD

**Issue by RaasAhsan** [typelevel/cats-effect#1029]
Going back and copying some of the ideas I had in https://typelevel.org/cats-effect/tracing/ to concrete issues. Since Cats Effect now supports fiber tracing, it goes without saying that we should support live fiber dumps as well. We should probably build out some of the other tracing features first before tackling this. 

Ideally the dump includes the following information:
1. The up-to-date trace for each fiber
2. Current carrier thread. This is related to #936.
3. Current fiber status. Some ideas are RUNNING, CANCELED, TERMINATED, ASYNC, WAITING
4. If we're in an asynchronous boundary, some auxiliary information reflecting what the asynchronous boundary is. e.g. if we're waiting on a `Deferred`, or if we're sleeping, etc.
5. Fiber identity and possibly ancestry

Some problems we should start thinking about:
1. Fiber runtimes may be managing thousands to hundreds of thousands of fibers at any given moment. Dumping all of them might be infeasible.
2. The act of producing a dump requires that we track all active running fibers. We should garbage collect those handles if the application no longer requires them, otherwise we are effectively creating a memory leak. `WeakReference` and `SoftReference` are of interest here.
3. How often should the manifest of running fibers be updated? I think the only reasonable answer here is at asynchronous boundaries.
4. How do we request a dump? Some ideas are registering a process signal handler, exposing a JMX MBean, or just a regular library API call.
5. Like other tracing features, the implementation of fiber dumps should have negligible performance impact when it isn't being used.

We should aim to get this in for Cats Effect 2, but think about how it would be designed for Cats Effect 3 as well.

**Comment by Daenyth** [comment]
> exposing a JMX MBean

I'd really like if it we had *at least* this, because when I get async locks, I often end up in visualvm trying to analyze a heap dump from the live system.

**Comment by djspiewak** [comment]
Okay I'm pretty sure this is now tractable. I have a PoC for very low overhead tracking of suspended fibers. Task list to get this finished up:

- [ ] Merge @armanbilge's stuff so we don’t create conflicts (#2207)
- [x] Clean up my PoC and PR
- [ ] Merge @vasilmkd's timer stuff so we don’t create conflicts (#2252)
- [ ] Get Vasil to PR his magic on live fiber enumeration
- [ ] Write a thingy which uses both functions (live and suspended) and gets all active fibers
- [ ] Use above thing to print all active traces (note: IOFiber needs an identifying toString)
- [ ] Wire up a signal handler hack so we get the above on Ctrl-\ (possibly using [`sun.misc.Signal`](https://github.com/frohoff/jdk8u-jdk/blob/master/src/share/classes/sun/misc/Signal.java) registering on `SIGQUIT`)

A lot of this can be done in parallel. The API for printing fiber dumps should be on `IORuntime`, while the signal handler should be on `IOApp`. Assume JVM-only for now but if we can get JS too that's awesome.

**Comment by armanbilge** [comment]
I think JS is tractable as well. We can copy in https://github.com/scala-js/scala-js-weakreferences/ under our own package and use a type alias to share the implementation. Then wherever the flag is to turn this on/off we also add a check if `WeakRef` is supported by the JS runtime. On Node.js we also have [signal handlers](https://nodejs.org/api/process.html#process_signal_events) and browsers will just have to use a global function or similar.

**Comment by vasilmkd** [comment]
JDK 17 most likely requires explicit `java` options for accessing `sun.misc.Signal`.

Edit: Rejoice. https://blogs.oracle.com/javamagazine/post/a-peek-into-java-17-continuing-the-drive-to-encapsulate-the-java-runtime-internals

Access to `sun.misc` classes continues to work.

**Comment by djspiewak** [comment]
This is now complete. Still some refining to go, particularly on WSL and on hypothetical platforms where `Signal` is unsupported, but we're in good shape in HEAD.
